Make sure that added R-groups have non-zero coordinates

ptosco opened the pull request:
Currently if cores have pre-defined 2D coordinates, e.g.
![image](https://user-images.githubusercontent.com/5244385/109877856-cef24800-7c73-11eb-8118-858a5df6db24.png)

added R-groups do not look good as they have zero coordinates:
![image](https://user-images.githubusercontent.com/5244385/109877775-b84bf100-7c73-11eb-8733-8096af20e3a9.png)

This PR fixes it:
![image](https://user-images.githubusercontent.com/5244385/109878146-2a243a80-7c74-11eb-8f9d-36322f1b8d90.png)

After I had already written a fix for this using `MolOps::addHs()`, I realized that #3767 already had code to do the same in a more straightforward way by exposing `setHydrogenCoords()`.
So I thought that rather than creating more review work with yet another RGD-related PR I'd rewrite what I had already written using code from #3767.
I volunteer to fix conflicts in #3767 after this is merged (they are very straightforward to fix).

bp-kelley commented:
This is good.  Part of me wonders if we should have an option or post-processing step to align the molecule to the drawn core.  This could make for easier visualizations.

bp-kelley commented:
@ptosco if you fix the conflicts, I can recheck this.

ptosco commented:
@bp-kelley Would you mind having a look at https://github.com/rdkit/rdkit/pull/3908 first, which I think is higher priority? Then I'll fix conflicts and merge this one. Thanks!

bp-kelley commented on Code/GraphMol/molopstest.cpp:
```diff
@@ -8051,6 +8051,38 @@ void testRingFamilies() {
 void testRingFamilies() {}
 #endif
 
+void testSetTerminalAtomCoords() {
```
What happens when you add two terminal atoms to the same atom?

greglandrum replied:
Should be fine. The AddHs() code uses that all the time.

ptosco commented:
@greglandrum @bp-kelley I have fixed the conflicts and added another test.
I have noticed that with current `master` something as innocent as this

```
import rdkit
from rdkit import Chem
from rdkit.Chem import rdRGroupDecomposition

core = Chem.MolFromSmiles("[1*]c1ccc([2*])cn1")
mol = Chem.MolFromSmiles("Fc1ccc([2*])cn1")
params = rdRGroupDecomposition.RGroupDecompositionParameters()
rgd = rdRGroupDecomposition.RGroupDecomposition(core, params)
rgd.Add(mol)
```

fails with

```
[22:29:48]

****
Pre-condition Violation
bad atom degree
Violation occurred on line 165 in file /scratch/toscopa1/src/rdkit/Code/GraphMol/AddHs.cpp
Failed Expression: mol->getAtomDegree(hydAtom) == 1
****

Traceback (most recent call last):
  File "/home/toscopa1/py/setHydrogenCoords.py", line 9, in <module>
    rgd.Add(mol)
RuntimeError: Pre-condition Violation
        bad atom degree
        Violation occurred on line 165 in file Code/GraphMol/AddHs.cpp
        Failed Expression: mol->getAtomDegree(hydAtom) == 1
        RDKIT: 2021.03.1dev1
        BOOST: 1_74
```

This PR also fixes this issue.
I have added a C++ test to cover the above case.
```

greglandrum commented on Code/GraphMol/RGroupDecomposition/testRGroupDecomp.cpp:
```diff
@@ -2115,6 +2115,85 @@ void testSingleAtomBridge() {
+      if (atom->getAtomicNum() == 0 && atom->getAtomMapNum() == 2) {
+        ++r2Num;
+        auto &r2Coord = rgdCore->getConformer().getAtomPos(atom->getIdx());
+        TEST_ASSERT(!(r2Coord.x > -1.e-4 && r2Coord.x < 1.e-4));
+        TEST_ASSERT(!(r2Coord.y > -1.e-4 && r2Coord.y < 1.e-4));
```
```suggestion
        TEST_ASSERT(fabs(r2Coord.x)>1e-4);
        TEST_ASSERT(fabs(r2Coord.y)>1e-4);
```

greglandrum commented on Code/GraphMol/RGroupDecomposition/testRGroupDecomp.cpp:
```diff
@@ -2115,6 +2115,85 @@ void testSingleAtomBridge() {
+          TEST_ASSERT(bond);
+          auto &nbrCoord = rgdCore->getConformer().getAtomPos(nbr->getIdx());
+          auto bondLen = (nbrCoord - r2Coord).length();
+          TEST_ASSERT(bondLen > 0.9 && bondLen < 1.1);
```
```suggestion
          TEST_ASSERT(fabs(bondLen - 1.0) < 0.1);
```

greglandrum commented on Code/GraphMol/RGroupDecomposition/testRGroupDecomp.cpp:
```diff
@@ -2115,6 +2115,85 @@ void testSingleAtomBridge() {
+        auto &r2Coord = rgdCore->getConformer().getAtomPos(atom->getIdx());
+        TEST_ASSERT(!(r2Coord.x > -1.e-4 && r2Coord.x < 1.e-4));
+        TEST_ASSERT(!(r2Coord.y > -1.e-4 && r2Coord.y < 1.e-4));
+        size_t nBonds = 0;
```
You don't need to count the number of bonds during the loop here.
The test you can do is very direct:
```
TEST_ASSERT(atom->getDegree()==1);

greglandrum commented on Code/GraphMol/MolOps.h:
```diff
@@ -192,6 +192,18 @@ RDKIT_GRAPHMOL_EXPORT void addHs(RWMol &mol, bool explicitOnly = false,
 
+//! Sets Cartesian coordinates for a terminal atom.
+//! Useful for growing an atom off a molecule with sensible
+//! coordinates based on the geometry of the neighbor.
+/*!
```
```suggestion
//! coordinates based on the geometry of the neighbor.
/*!
    NOTE: this sets appropriate coordinates in all of the molecule's conformers.
```

greglandrum commented on Code/GraphMol/Wrap/MolOps.cpp:
```diff
@@ -1012,6 +1012,24 @@ struct molops_wrapper {
+        "Sets Cartesian coordinates for a terminal atom.\n\
+\n\
+  Useful for growing an atom off a molecule with sensible \n\
+  coordinates based on the geometry of the neighbor.\n\
+\n\
```
```suggestion
  coordinates based on the geometry of the neighbor.\n\
\n\
  NOTE: this sets the appropriate coordinates in all of the molecule's conformers \n\
```

greglandrum reviewed (changes requested):
small suggestions

ptosco commented:
@greglandrum Thanks Greg, I have applied all your suggested changes, they all make sense.

greglandrum commented:
I fixed the merge conflicts. Will merge after the CI builds pass
